libpod/image/image.go

```diff
@@ -1076,3 +1078,144 @@ func (i *Image) Comment(ctx context.Context, manifestType string) (string, error
 	}
 	return ociv1Img.History[0].Comment, nil
 }
+
+type TreeImage struct {
+	ID       string
+	ParentID string
+	RepoTags []string
+	Size     int64
+}
+
+// Tree gets dependency list in tree format of the image
+func (i *Image) Tree(ctx context.Context, imageName string) (*bytes.Buffer, error) {
+
+	layerMap, err := i.layerToImageMap(ctx)
+	if err != nil {
+		return nil, err
+	}
+	layerMap, err = i.addImageToLayerInfo(ctx, layerMap)
+
+	repoTag, err := i.MatchRepoTag(imageName)
+	if err != nil {
+		return nil, err
+	}
+	// Get the image top layer and then parse as per parent
+	layer := i.TopLayer()
+	dependencyList := returnDependencyTree(layerMap, repoTag, layer)
+
+	var buffer bytes.Buffer
+	printTree(&buffer, dependencyList)
+
+	return &buffer, err
+}
+
+func printTree(buffer *bytes.Buffer, treeStack []TreeImage) {
+
+	prefixSpace := 0
+	for i := len(treeStack) - 1; i >= 0; i-- {
+		for p := 0; p < prefixSpace; p++ {
+			buffer.WriteString(fmt.Sprintf(" "))
+		}
+		buffer.WriteString(fmt.Sprintf("└─ ID: %s  Tags: %s VirtualSize: %v\n",
+			treeStack[i].ID[:12],
+			treeStack[i].RepoTags,
+			units.HumanSizeWithPrecision(float64(treeStack[i].Size), 4)),
+		)
+
+		prefixSpace++
+	}
+}
+
+func (i *Image) layerToImageMap(ctx context.Context) (map[string]*TreeImage, error) {
+	layerMap := make(map[string]*TreeImage)
+	layers, _ := i.imageruntime.store.Layers()
+	for _, layer := range layers {
+		existingLayer, ok := layerMap[layer.ID]
+		if !ok {
+			layerMap[layer.ID] = &TreeImage{
+				ID:       layer.ID,
+				Size:     layer.CompressedSize,
+				ParentID: layer.Parent,
+			}
+		} else {
+			return nil, fmt.Errorf("duplicate %s layer found in image store", existingLayer.ID)
+		}
+	}
+	return layerMap, nil
+}
+
+func (i *Image) addImageToLayerInfo(ctx context.Context, layerMap map[string]*TreeImage) (map[string]*TreeImage, error) {
+	images, err := i.imageruntime.GetImages()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting images from store")
+	}
+
+	for _, image := range images {
+
+		layerid := image.TopLayer()
+		imgLayer, err := image.imageruntime.store.Layer(layerid)
+		if err != nil {
+			return nil, err
+		}
+
+		for imgLayer != nil {
+			existingLayer, ok := layerMap[imgLayer.ID]
+			if !ok {
+				return nil, fmt.Errorf("layer %s not found ", imgLayer.ID)
+			} else {
+				names := image.Names()
+				existingLayer.RepoTags = append(existingLayer.RepoTags, names...)
+			}
+			imgLayer = image.getLayer(imgLayer.Parent)
```

Comment:
> Can we extract information of all layer to their corresponding images without parsing all Images ? NO
> To map the layer with image in O(1), we need this mapping.

(Well, that, _again_, goes back to what is the intended output format and semantics…)

> shall be helpful, if later multiple images are given as input to tree command.

Do you mean representing something like
```
└─ fedora
 └─ (unnamed-shared)
  └─ (unnamed-app1-only)
   └─ app1
  └─ (unnamed-app2-only)
   └─ app2
  └─ (unnamed-app3-only)
   └─ app3
```
and perhaps that `podman image tree app1 app2` should exclude the `app3` image, so building the full `RepoTags` arrays as they are now is useful for filtering?

Even in that case:
- It’s not _obviously_ the right thing to output the "app1 app2 app3" tag names for the “(unnamed-shared)” and `fedora` layers; the information is already conveyed by the shape, and repeating the names may be just unwanted clutter for human readers. (But it’s also not obviously nonsensical — what is the intent?)
- Filtering the tree (so that when the user asks for app1 app2, the app3 subtree is skipped)  does not need the names propagated to be efficient, and in fact is _less_ efficient that way. Using the labels for filtering means something like
  ```
  filtered := emptySet()
  for layer in layers:
    for tag in layer.RepoTags:
      if CLIArgsSet.contains(tag):
        filtered.add(layer)
  ```
  which is O(layers*tags), but we can instead do:
  ```
  filtered := emptySet()
  queue := emptyQueue
  for tag in CLIArgsSet:
    queue.push(topLevelLayerByTag[tag])
  while layer := queue.pop():
    filtered.add(layer)
    if layer.Parent != "" && !filtered.contains(layer.Parent)
     queue.push(layer.Parent)
  ```
  which is O(layers+tags).
>> is really a presentation matter — it may make sense to let the user ask what images use a specific layer, but it’s not obviously reasonable to list the top-level rhel image layer with dozens of names.
>
> Still only one name is printed as output. Multiple names are just part of intermediate information.

Yes, the code only prints one name, but _which one_? Right now it is essentially random (depending on the order the images are encountered), but the user would almost always want to know the tags that refer to this layer as the top layer of an image (which may be more than one value!) primarily.

Reply:
> which is O(layers*tags)

(Note that this does not get better if `layer.RepoTags` becomes a set with O(1) access instead of the current array - intersecting the `layer.RepoTags` set with the hypothetical `CLIArgsSet` would still be O(relevant tags).

And, sure, this may all be premature optimization at our current scales — it’s just that propagating the names through all the children is not making that premature optimization any easier AFAICS.)